pkg/controllers/cluster/kubeconfig_controller.go

```diff
@@ -0,0 +1,257 @@
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=greenhouse.sap,resources=cluster-kubeconfigs,verbs=get;list;watch;create;update;patch
+
+// SetupWithManager sets up the controller with the Manager
+func (r *KubeconfigReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		Watches(&v1alpha1.Cluster{}, handler.EnqueueRequestsFromMapFunc(sameNameResource)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(clusterSecretToCluster)).
+		Watches(&v1alpha1.Organization{}, handler.EnqueueRequestsFromMapFunc(r.organizationToClusters)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(r.organizationSecretToClusters)).
+		For(&v1alpha1.ClusterKubeconfig{}).
+		Complete(r)
+}
+
+func (r *KubeconfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithValues("cluster", req.Name, "namespace", req.Namespace)
+
+	var cluster v1alpha1.Cluster
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+		l.Info("skip reconcile, cluster is not found")
+		return ctrl.Result{}, nil
+	}
+
+	if cluster.GetDeletionTimestamp() != nil {
+		l.Info("skip reconcile, cluster is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	var kubeconfig v1alpha1.ClusterKubeconfig
+
+	if err := r.Get(ctx, req.NamespacedName, &kubeconfig); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+
+		l.Info("kubeconfig not found, will be created")
+		kubeconfig.Name = cluster.Name
+		kubeconfig.Namespace = cluster.Namespace
+		kubeconfig.Spec.Kubeconfig = v1alpha1.ClusterKubeconfigData{
+			Kind:       "Config",
+			APIVersion: "v1",
+		}
+		kubeconfig.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: cluster.APIVersion,
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			},
+		}
+		kubeconfig.Spec.Kubeconfig.Contexts = []v1alpha1.ClusterKubeconfigContextItem{
+			{
+				Name: cluster.Name,
+				Context: v1alpha1.ClusterKubeconfigContext{
+					Cluster:   cluster.Name,
+					AuthInfo:  "oidc@" + cluster.Name,
+					Namespace: "default",
+				},
+			}}
+		kubeconfig.Spec.Kubeconfig.CurrentContext = cluster.Name
+		kubeconfig.Status.Conditions.SetConditions(v1alpha1.TrueCondition(v1alpha1.KubeconfigCreatedCondition, "NewCreation", ""))
+	}
+
+	// get oidc info from organization
+	oidc, err := r.getOIDCInfo(ctx, cluster.Namespace)
+	if err != nil {
+		l.Error(err, "skip reconcile, oidc data not fetched")
+		return ctrl.Result{}, nil
+	}
+
+	// get cluster connection data from cluster secret
+	var secret corev1.Secret
+	err = r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, &secret)
+	if err != nil {
```

Comment:
Imo this should be reflected into an Condition. If we are only treating Clusters with StatusCondition `Ready`, no error should happen here. But if it still happens we could persist into `ClusterReady` on the Kubeconfig as well.

Reply:
All errors are persisted under Conditions by 16b7fb7af545499f90a2f168752ee2ceb335a60a